### PR TITLE
[SCB-1577] Modify the OnConnected method to bidirectional streaming RPC

### PR DESCRIPTION
**Pull request by coolbeevip**
This is the first step to Improve the transaction compensation mechanism for FSM

[1] https://issues.apache.org/jira/browse/SCB-1622

### CODE REVIEW

File: alpha/alpha-server/src/main/java/org/apache/servicecomb/pack/alpha/server/GrpcTxEventEndpointImpl.java [hunk 1]
```diff
@@ -56,10 +60,29 @@
...
+      }
+
+      @Override
+      public void onError(Throwable throwable) {
+        LOG.error(throwable.getMessage());
+      }
+
+      @Override
+      public void onCompleted() {
+        // Do nothing here
```
Comment (WillemJiang) [inline alpha/alpha-server/src/main/java/org/apache/servicecomb/pack/alpha/server/GrpcTxEventEndpointImpl.java]: I think we need to add some log here.

File: omega/omega-connector/omega-connector-grpc/src/main/java/org/apache/servicecomb/pack/omega/connector/grpc/core/ReconnectStreamObserver.java [hunk 2]
```diff
@@ -42,10 +45,23 @@ public ReconnectStreamObserver(
...
   @Override
   public void onCompleted() {
     // Do nothing here
   }
+
+  public void cancelWait(){
+    latch.countDown();
+  }
+
+  public void waitConnected() {
```
Comment (WillemJiang) [inline omega/omega-connector/omega-connector-grpc/src/main/java/org/apache/servicecomb/pack/omega/connector/grpc/core/ReconnectStreamObserver.java]: When will you call the waitConnected?
I don't think it's a good practice to introduce sync wait here.

File: omega/omega-connector/omega-connector-grpc/src/main/java/org/apache/servicecomb/pack/omega/connector/grpc/saga/GrpcSagaClientMessageSender.java [hunk 3]
```diff
@@ -68,7 +73,8 @@ public GrpcSagaClientMessageSender(
 
   @Override
   public void onConnected() {
-    asyncEventService.onConnected(serviceConfig, compensateStreamObserver);
+    asyncEventService.onConnected(compensateStreamObserver).onNext(serviceConfig);
+    compensateStreamObserver.waitConnected();
```
Comment (coolbeevip) [inline omega/omega-connector/omega-connector-grpc/src/main/java/org/apache/servicecomb/pack/omega/connector/grpc/saga/GrpcSagaClientMessageSender.java]: Waiting for Alpha to send a connection response

File: omega/omega-connector/omega-connector-grpc/src/main/java/org/apache/servicecomb/pack/omega/connector/grpc/core/ReconnectStreamObserver.java (hunk 2 above)
Comment (coolbeevip) [reply omega/omega-connector/omega-connector-grpc/src/main/java/org/apache/servicecomb/pack/omega/connector/grpc/core/ReconnectStreamObserver.java]: > When will you call the waitConnected?

Waiting only happens when Omega first connects to Alpha. 
I marked the location where call waitConnected in GrpcSagaClientMessageSender.java L77

> I don't think it's a good practice to introduce sync wait here.

The original idea was when Alpha receives the connection request, it sends a response message to Omega. Omega receives the response message and then continues.

if you don't feel it's needed  I would remove them

File: alpha/alpha-server/src/main/java/org/apache/servicecomb/pack/alpha/server/fsm/GrpcSagaEventService.java [hunk 4]
```diff
@@ -54,11 +54,32 @@ public GrpcSagaEventService(ActorEventChannel actorEventChannel,
...
+
+      @Override
+      public void onNext(GrpcServiceConfig grpcServiceConfig) {
+        grpcOmegaCallback = new GrpcOmegaCallback(responseObserver);
+        omegaCallbacks
+            .computeIfAbsent(grpcServiceConfig.getServiceName(), key -> new ConcurrentHashMap<>())
+            .put(grpcServiceConfig.getInstanceId(), grpcOmegaCallback);
+        responseObserver.onNext(GrpcCompensateCommand.newBuilder()
+            .setConnectedResponse(true)
+            .build());
```
Comment (coolbeevip) [inline alpha/alpha-server/src/main/java/org/apache/servicecomb/pack/alpha/server/fsm/GrpcSagaEventService.java]: @WillemJiang Alpha sends a response message to Omega after receiving the connection request.

File: alpha/alpha-server/src/main/java/org/apache/servicecomb/pack/alpha/server/GrpcTxEventEndpointImpl.java (hunk 1 above)
Comment (coolbeevip) [reply alpha/alpha-server/src/main/java/org/apache/servicecomb/pack/alpha/server/GrpcTxEventEndpointImpl.java]: done

File: omega/omega-connector/omega-connector-grpc/src/main/java/org/apache/servicecomb/pack/omega/connector/grpc/core/ReconnectStreamObserver.java (hunk 2 above)
Comment (coolbeevip) [reply omega/omega-connector/omega-connector-grpc/src/main/java/org/apache/servicecomb/pack/omega/connector/grpc/core/ReconnectStreamObserver.java]: done

**Comment by WillemJiang** [comment]
@coolbeevip This PR is good for me.
I think we can merge it,  do you mind remove the WIP perfix？You can send other PR for the implementation of next steps.

**Comment by coolbeevip** [comment]
> @coolbeevip This PR is good for me.
> I think we can merge it, do you mind remove the WIP perfix？You can send other PR for the implementation of next steps.

OK, I will remove the WIP prefix and split the issue of JIRA

**Comment by coolbeevip** [comment]
I created a new JIRA[1] and convert SCB-1577 into its sub-task

[1] https://issues.apache.org/jira/browse/SCB-1622

**Pull request merged**